src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -44,16 +44,41 @@ public override BoundNode VisitForEachStatement(BoundForEachStatement node)
                     return RewriteMultiDimensionalArrayForEachStatement(node);
                 }
             }
-            else if (nodeExpressionType.SpecialType == SpecialType.System_String)
+            else if (CanRewriteForEachAsFor(node.Syntax, nodeExpressionType, out var indexerGet, out var lengthGetter))
             {
-                return RewriteStringForEachStatement(node);
+                return RewriteForEachStatementAsFor(node, indexerGet, lengthGetter);
             }
             else
             {
                 return RewriteEnumeratorForEachStatement(node);
             }
         }
 
+        private bool CanRewriteForEachAsFor(SyntaxNode forEachSyntax, TypeSymbol nodeExpressionType, out MethodSymbol indexerGet, out MethodSymbol lengthGet)
+        {
+            lengthGet = indexerGet = null;
+            var origDefinition = nodeExpressionType.OriginalDefinition;
+
+            if (origDefinition.SpecialType == SpecialType.System_String)
+            {
+                lengthGet = UnsafeGetSpecialTypeMethod(forEachSyntax, SpecialMember.System_String__Length);
+                indexerGet = UnsafeGetSpecialTypeMethod(forEachSyntax, SpecialMember.System_String__Chars);
+            }
+            else if ((object)origDefinition == this._compilation.GetWellKnownType(WellKnownType.System_Span_T))
+            {
+                var spanType = (NamedTypeSymbol)nodeExpressionType;
+                lengthGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Span_T__get_Length, isOptional: true)?.SymbolAsMember(spanType);
+                indexerGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Span_T__get_Item, isOptional: true)?.SymbolAsMember(spanType);
+            }
+            else if ((object)origDefinition == this._compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T))
+            {
+                var spanType = (NamedTypeSymbol)nodeExpressionType;
+                lengthGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_ReadOnlySpan_T__get_Length, isOptional: true)?.SymbolAsMember(spanType);
+                indexerGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_ReadOnlySpan_T__get_Item, isOptional: true)?.SymbolAsMember(spanType);
+            }
+            return lengthGet != null && indexerGet != null;
```

Comment:
It seems that the behavior for malformed `Span` type is to fall back (lower foreach with enumerator instead of using a for loop).
Is that desirable? I'd think this should give an error.

Reply:
Why would that be better? Spec says as long as a type implements foreach pattern, it should work with foreach. 
We can emit it as 'for' as a matter of optimization, but in an unlikely case that we can't, we should emit as `foreach`

---
In reply to: [153595276](https://github.com/dotnet/roslyn/pull/23363#discussion_r153595276) [](ancestors = 153595276)

Reply:
It could also be breaking ReadOnlySpan that has byval indexer. Which works fine now and is reasonably fast (although indexing is slightly better)

---
In reply to: [153607893](https://github.com/dotnet/roslyn/pull/23363#discussion_r153607893) [](ancestors = 153607893,153595276)